tests/InputValidationMiddlewareTest.php

```diff
@@ -0,0 +1,189 @@
+<?php
+namespace Aws\Test;
+
+use Aws\Api\DateTimeResult;
+use Aws\AwsClient;
+use Aws\EndpointParameterMiddleware;
+use Aws\HandlerList;
+use Aws\Api\Service;
+use Aws\InputValidationMiddleware;
+use Cassandra\Time;
+use GuzzleHttp\Psr7\Request;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @covers \Aws\EndpointParameterMiddleware
+ */
+class InputValidationMiddlewareTest extends TestCase
+{
+    /**
+     * Data provider for exceptions treated as invalid argument exceptions
+     *
+     * @return array
+     */
+    public function getInvalidEndpointExceptions()
+    {
+        return [
+            [''],
+            [' '],
+            ['  '],
+            [null],
+        ];
+    }
+
+    /**
+     * Data provider for exceptions treated as invalid argument exceptions
+     *
+     * @return array
+     */
+    public function getValidInputs()
+    {
+        return [
+            ['existing data'],
+            [' q '],
+            [[]],
+            [['abc']],
+            [0],
+            [DateTimeResult::fromEpoch(time())],
+        ];
+    }
+
+    /**
+     * @dataProvider getInvalidEndpointExceptions
+     *
+     * @param $input
+     */
+    public function testThrowsExceptions($input)
+    {
+        $service = $this->generateTestService();
+        $client = $this->generateTestClient($service);
+        $command = $client->getCommand('RequiredOp', ['InputParameter' => $input]);
+        $mandatoryInputList = ['InputParameter'];
+
+        $list = new HandlerList();
+        $list->setHandler(function ($command) {
+            return;
+        });
+        $list->appendValidate(
+            InputValidationMiddleware::wrap($service, $mandatoryInputList)
+        );
+        $handler = $list->resolve();
+
+        try {
+            $handler($command, new Request('POST', 'https://foo.com'));
+            $this->fail('Test should have thrown an InvalidArgumentException for not having required parameter set.');
+        } catch (\InvalidArgumentException $e) {
+            $this->assertEquals(
+                "The RequiredOp operation requires non-empty parameter: InputParameter",
+                $e->getMessage()
+            );
+        }
+    }
+
+    /**
+     * @dataProvider getInvalidEndpointExceptions
+     *
+     * @param $input
+     */
+    public function testNoValidationWithoutInputLIst($input)
+    {
+        $service = $this->generateTestService();
+        $client = $this->generateTestClient($service);
+        $command = $client->getCommand('RequiredOp', ['InputParameter' => $input]);
+        $mandatoryInputList = [];
+        $list = new HandlerList();
+        $list->setHandler(function ($command) {
+            return;
+        });
+        $list->appendValidate(
+            InputValidationMiddleware::wrap($service, $mandatoryInputList)
+        );
+        $handler = $list->resolve();
+
+        $handler($command, new Request('POST', 'https://foo.com'));
+        self::assertTrue(true);
```

Comment:
Why is this trivial assert here? Perhaps a slightly more meaningful assert would be capturing the result of the $handler() and checking that it's going through the stack as expected? You may have to modify the defined handler to return something useful.